In [1]:
import numpy as np
import pandas as pd
import os

In [2]:
ords_prods_merged = pd.read_pickle(r'C:\Users\katey\Desktop\02 Data\Prepared Data\orders_products_merged_updated.pk1')

In [3]:
df_dept = pd.read_csv(r'C:\Users\katey\Desktop\02 Data\Prepared Data\departments_wrangled.xlsx')

In [4]:
df_dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [5]:
ords_prods_merged.shape

(13060458, 21)

In [6]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             2    473747        1    prior             3          3   
1             4    431534        1    prior             5          4   
2             6    550135        1    prior             7          1   
3           151    487368       15    prior            22          1   
4           214    532817       19    prior             7          4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                 12                    21.0       196.0                1.0   
1                 15                    28.0       196.0                1.0   
2                  9                    20.0       196.0                1.0   
3                 10                    14.0       196.0                1.0   
4                 17                     6.0       196.0                1.0   

   ...  Unnamed: 0_y  product_name aisle_id  department_id  prices  _merge  \
0  ...           195          Soda       77              7     9.0    both   
1  ...           195          Soda       77              7     9.0    both   
2  ...           195          Soda       77              7     9.0    both   
3  ...           195          Soda       77              7     9.0    both   
4  ...           195          Soda       77              7     9.0    both   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Least busy days           Most orders  
1  Mid-range product      Least busy  Least busy days           Most orders  
2  Mid-range product  Regularly busy     Busiest days           Most orders  
3  Mid-range product  Regularly busy     Busiest days           Most orders  
4  Mid-range product      Least busy  Least busy days        Average orders  

[5 rows x 21 columns]

In [8]:
ords_prods_merged.columns

Index(['Unnamed: 0_x', 'order_id', 'user_id', 'eval_set', 'order_number',
       'order_dow', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', 'Unnamed: 0_y',
       'product_name', 'aisle_id', 'department_id', 'prices', '_merge',
       'price_range_loc', 'busiest_day', 'busiest_days',
       'busiest_period_of_day'],
      dtype='object')

In [9]:
ords_prods_merged.groupby('product_name')

In [10]:
ords_prods_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.427839
2                17.242992
3                17.176399
4                17.828535
5                15.225470
6                16.430332
7                17.236324
8                15.433071
9                15.949204
10               20.540360
11               16.162394
12               15.904644
13               16.589470
14               16.824487
15               16.202673
16               17.656786
17               15.732946
18               19.354771
19               17.153130
20               16.464613
21               22.917563

The minimum mean is 15.22 for department 5 and maxium mean is 22.92 for department 21. There is less variability of the mean if a subset is used rather than the entire database that has missing department labels.

In [11]:
ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              2    473747        1    prior             3          3   
1              4    431534        1    prior             5          4   
2              6    550135        1    prior             7          1   
3            151    487368       15    prior            22          1   
4            214    532817       19    prior             7          4   
..           ...       ...      ...      ...           ...        ...   
95         27052   1210848     1680    prior            33          1   
96         27055    486609     1680    prior            36          3   
97         27056    112979     1680    prior            37          5   
98         27791    231329     1723    prior             2          1   
99         27839    835467     1726    prior             2          5   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  12                    21.0       196.0                1.0   
1                  15                    28.0       196.0                1.0   
2                   9                    20.0       196.0                1.0   
3                  10                    14.0       196.0                1.0   
4                  17                     6.0       196.0                1.0   
..                ...                     ...         ...                ...   
95                 13                     6.0       196.0                1.0   
96                 18                     6.0       196.0                1.0   
97                 14                    23.0       196.0                2.0   
98                  9                    30.0       196.0                3.0   
99                 18                    17.0       196.0                2.0   

    ...  product_name  aisle_id department_id  prices  _merge  \
0   ...          Soda        77             7     9.0    both   
1   ...          Soda        77             7     9.0    both   
2   ...          Soda        77             7     9.0    both   
3   ...          Soda        77             7     9.0    both   
4   ...          Soda        77             7     9.0    both   
..  ...           ...       ...           ...     ...     ...   
95  ...          Soda        77             7     9.0    both   
96  ...          Soda        77             7     9.0    both   
97  ...          Soda        77             7     9.0    both   
98  ...          Soda        77             7     9.0    both   
99  ...          Soda        77             7     9.0    both   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Least busy days           Most orders   
1   Mid-range product      Least busy  Least busy days           Most orders   
2   Mid-range product  Regularly busy     Busiest days           Most orders   
3   Mid-range product  Regularly busy     Busiest days           Most orders   
4   Mid-range product      Least busy  Least busy days        Average orders   
..                ...             ...              ...                   ...   
95  Mid-range product  Regularly busy     Busiest days           Most orders   
96  Mid-range product  Regularly busy  Least busy days        Average orders   
97  Mid-range product  Regularly busy   Regularly busy           Most orders   
98  Mid-range product  Regularly busy     Busiest days           Most orders   
99  Mid-range product  Regularly busy   Regularly busy        Average orders   

   max_order  
0          7  
1          7  
2          7  
3         22  
4          9  
..       ...  
95        41  
96        41  
97        41  
98         3  
99         7  

[100 rows x 22 columns]

In [13]:
pd.options.display.max_rows = None

In [14]:
ords_prods_merged.loc[ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [15]:
ords_prods_merged.loc[(ords_prods_merged['max_order'] <= 40) & (ords_prods_merged['max_order'] >= 10), 'loyalty_flag'] = 'Regular Customer'

In [16]:
ords_prods_merged.loc[ords_prods_merged['max_order'] < 10, 'loyalty_flag'] = 'New Customer'

In [17]:
ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

Regular Customer    6541620
Loyal Customer      3960839
New Customer        2557999
Name: loyalty_flag, dtype: int64

In [18]:
ords_prods_merged.groupby('prices')

In [19]:
ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal Customer    10.539071
New Customer      13.483071
Regular Customer  12.709585

In [20]:
ords_prods_merged['mean_price'] = ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [21]:
ords_prods_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              2    473747        1    prior             3          3   
1              4    431534        1    prior             5          4   
2              6    550135        1    prior             7          1   
3            151    487368       15    prior            22          1   
4            214    532817       19    prior             7          4   
5            752   1318871       52    prior             2          1   
6            753   1261384       52    prior             3          1   
7            755    580568       52    prior             5          2   
8            759    342306       52    prior             9          1   
9            762    944694       52    prior            12          1   
10           765   1268191       52    prior            15          2   
11           767   1122089       52    prior            17          2   
12           769    180919       52    prior            19          3   
13          1023    970922       67    prior             2          1   
14          1024    596585       67    prior             3          1   
15          1035    551869       67    prior            14          1   
16          1037    768572       67    prior            16          4   
17          1039    150528       67    prior            18          5   
18          1040     29758       67    prior            19          0   
19          1043    863974       67    prior            22          1   
20          1210    499460       82    prior             3          5   
21          1216   1358880       82    prior             9          1   
22          1217    343588       82    prior            10          2   
23          1225    147262       82    prior            18          1   
24          1486    681011       98    prior             9          3   
25          1489    747431       98    prior            12          6   
26          1744    520620      120    prior             1          3   
27          2909    276171      195    prior            19          5   
28          2930    417035      195    prior            40          3   
29          2947    745908      195    prior            57          4   
30          6740    858092      420    prior             4          1   
31          6741    119314      420    prior             5          3   
32          9333    668870      591    prior             2          1   
33          9839    981587      626    prior            12          1   
34          9847    716308      626    prior            20          3   
35          9912    325835      626    prior            85          1   
36         10136    490788      642    prior             2          2   
37         10875     46887      676    prior             6          0   
38         10877   1172680      676    prior             8          0   
39         10879    443065      676    prior            10          5   
40         10880   1171861      676    prior            11          2   
41         10923    296844      680    prior             5          5   
42         11196    962470      691    prior            14          6   
43         11202   1060084      691    prior            20          1   
44         11435    147387      709    prior             4          0   
45         12675    231498      793    prior             1          1   
46         12676    657712      793    prior             2          2   
47         12679     79610      793    prior             5          3   
48         15971   1121647      992    prior             7          5   
49         15973    212048      992    prior             9          0   
50         15974    681310      992    prior            10          5   
51         15975     58766      992    prior            11          5   
52         15977   1238233      992    prior            13          6   
53         15980     88614      992    prior            16    

In [22]:
ords_prods_merged.loc[ords_prods_merged['mean_price'] >= 10, 'spending_flag'] = 'High Spender'

In [23]:
ords_prods_merged.loc[ords_prods_merged['mean_price'] < 10, 'spending_flag'] = 'Low Spender'

In [24]:
ords_prods_merged['spending_flag'].value_counts(dropna = False)

Low Spender     12821465
High Spender      238993
Name: spending_flag, dtype: int64

In [25]:
ords_prods_merged['median_days'] = ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
ords_prods_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              2    473747        1    prior             3          3   
1              4    431534        1    prior             5          4   
2              6    550135        1    prior             7          1   
3            151    487368       15    prior            22          1   
4            214    532817       19    prior             7          4   
5            752   1318871       52    prior             2          1   
6            753   1261384       52    prior             3          1   
7            755    580568       52    prior             5          2   
8            759    342306       52    prior             9          1   
9            762    944694       52    prior            12          1   
10           765   1268191       52    prior            15          2   
11           767   1122089       52    prior            17          2   
12           769    180919       52    prior            19          3   
13          1023    970922       67    prior             2          1   
14          1024    596585       67    prior             3          1   
15          1035    551869       67    prior            14          1   
16          1037    768572       67    prior            16          4   
17          1039    150528       67    prior            18          5   
18          1040     29758       67    prior            19          0   
19          1043    863974       67    prior            22          1   
20          1210    499460       82    prior             3          5   
21          1216   1358880       82    prior             9          1   
22          1217    343588       82    prior            10          2   
23          1225    147262       82    prior            18          1   
24          1486    681011       98    prior             9          3   
25          1489    747431       98    prior            12          6   
26          1744    520620      120    prior             1          3   
27          2909    276171      195    prior            19          5   
28          2930    417035      195    prior            40          3   
29          2947    745908      195    prior            57          4   
30          6740    858092      420    prior             4          1   
31          6741    119314      420    prior             5          3   
32          9333    668870      591    prior             2          1   
33          9839    981587      626    prior            12          1   
34          9847    716308      626    prior            20          3   
35          9912    325835      626    prior            85          1   
36         10136    490788      642    prior             2          2   
37         10875     46887      676    prior             6          0   
38         10877   1172680      676    prior             8          0   
39         10879    443065      676    prior            10          5   
40         10880   1171861      676    prior            11          2   
41         10923    296844      680    prior             5          5   
42         11196    962470      691    prior            14          6   
43         11202   1060084      691    prior            20          1   
44         11435    147387      709    prior             4          0   
45         12675    231498      793    prior             1          1   
46         12676    657712      793    prior             2          2   
47         12679     79610      793    prior             5          3   
48         15971   1121647      992    prior             7          5   
49         15973    212048      992    prior             9          0   
50         15974    681310      992    prior            10          5   
51         15975     58766      992    prior            11          5   
52         15977   1238233      992    prior            13          6   
53         15980     88614      992    prior            16    

In [27]:
ords_prods_merged.loc[ords_prods_merged['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [28]:
ords_prods_merged.loc[(ords_prods_merged['median_days'] <= 20) & (ords_prods_merged['median_days'] > 10), 'frequency_flag'] = 'Regular customer'

In [29]:
ords_prods_merged.loc[ords_prods_merged['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [30]:
ords_prods_merged['frequency_flag'].value_counts(dropna=False)

Frequent customer        8663110
Regular customer         2708070
Non-frequent customer    1623330
NaN                        65948
Name: frequency_flag, dtype: int64

In [31]:
ords_prods_merged['frequency_flag'].fillna("Non-frequent customer", inplace = True)

In [32]:
ords_prods_merged['frequency_flag'].value_counts(dropna=False)

Frequent customer        8663110
Regular customer         2708070
Non-frequent customer    1689278
Name: frequency_flag, dtype: int64